readthedocs/oauth/services/githubapp.py

```diff
@@ -0,0 +1,473 @@
+        - https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-app-installations-accessible-to-the-user-access-token
+
+        .. note::
+
+           If the installation wasn't in our database, we create it
+           (but we don't sync the repositories, since the caller should be responsible for that).
+
+        .. note::
+
+           User access tokens expire after 8 hours, but our OAuth2 client should handle refreshing the token.
+           But, the refresh token expires after 6 months, in order to refresh that token,
+           the user needs to sign in using GitHub again (just a normal sing-in, not a re-authorization or sign-up).
+        """
+        social_accounts = SocialAccount.objects.filter(
+            user=user,
+            provider=cls.allauth_provider.id,
+        )
+        for account in social_accounts:
+            oauth2_client = get_oauth2_client(account)
+            resp = oauth2_client.get("https://api.github.com/user/installations")
+            if resp.status_code != 200:
+                log.info(
+                    "Failed to fetch installations from GitHub",
+                    user=user,
+                    account_id=account.uid,
+                    status_code=resp.status_code,
+                    response=resp.json(),
+                )
+                continue
+
+            for gh_installation in resp.json()["installations"]:
+                (
+                    installation,
+                    _,
+                ) = GitHubAppInstallation.objects.get_or_create_installation(
+                    installation_id=gh_installation["id"],
+                    target_id=gh_installation["target_id"],
+                    target_type=gh_installation["target_type"],
+                    extra_data={"installation": gh_installation},
+                )
+                yield cls(installation)
+
+    @classmethod
+    def sync_user_access(cls, user):
+        """
+        Sync the user's access to the provider repositories and organizations.
+
+        Since we are using a GitHub App, we don't have a way to check all the repositories and organizations
+        the user has access to or if it lost access to a repository or organization.
+
+        Our webhooks should keep permissions in sync, but just in case,
+        we first sync the repositories from all installations accessible to the user (refresh access to new repositories),
+        and then we sync each repository the user has access to (check if the user lost access to a repository, or his access level changed).
+        """
+        has_error = False
+        # Refresh access to all installations accessible to the user.
+        for service in cls.for_user(user):
+            try:
+                service.sync()
+            except SyncServiceError:
+                # Don't stop the sync if one installation fails,
+                # as we should try to sync all installations.
+                has_error = True
```

Comment:
We're not capturing the error here -- are we expecting `sync` to log it or something?